Add temporary backdrop workaround for fullscreen elements

nt1m opened the pull request:
#### d7bab9d6f7c45b99da17d78a464eac1fedff0ead
<pre>
Add temporary backdrop workaround for fullscreen elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=247349">https://bugs.webkit.org/show_bug.cgi?id=247349</a>
rdar://101600873

Reviewed by Tim Horton.

Removing RenderFullScreen introduced a regression, where content appears behind the fullscreen element.
RenderFullScreen used to solve that by wrapping the fullscreen element renderer with a renderer that has a black background.

The standard way of solving this issue is to add a ::backdrop that has such a background. This will be done once fullscreen
switches to use the top layer.

For now, add a pseudo-class that targets the parent of the fullscreen element, and use ::before to emulate the ::backdrop.
We specifically pick the parent, because it is guaranteed to be able to contain children (and ::before), and to avoid conflict
with any potential ::before on the fullscreen element itself.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesFullScreenParentPseudoClass):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/css/fullscreen.css:
(:-webkit-full-screen-parent::before):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:

Canonical link: <a href="https://commits.webkit.org/256226@main">https://commits.webkit.org/256226@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/036a405146320c809a0e14261a5680313fce6ab3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95132 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4272 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/28156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104729 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164980 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4369 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33109 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87445 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100626 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100800 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3183 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81669 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30161 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/86925 "Build was cancelled. Recent messages:") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73056 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38862 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36682 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19762 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4305 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40617 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39003 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -425,6 +425,14 @@ ALWAYS_INLINE bool matchesFullScreenAnimatingFullScreenTransitionPseudoClass(con
     return element.document().fullscreenManager().isAnimatingFullscreen();
 }
 
+/* FIXME: Remove this temporary hack */
```
This should reference a bugzilla bug (and probably say why instead of calling it a hack)

nt1m committed:
Add temporary backdrop workaround for fullscreen elements
https://bugs.webkit.org/show_bug.cgi?id=247349
rdar://101600873

Reviewed by Tim Horton.

Removing RenderFullScreen introduced a regression, where content appears behind the fullscreen element.
RenderFullScreen used to solve that by wrapping the fullscreen element renderer with a renderer that has a black background.

The standard way of solving this issue is to add a ::backdrop that has such a background. This will be done once fullscreen
switches to use the top layer.

For now, add a pseudo-class that targets the parent of the fullscreen element, and use ::before to emulate the ::backdrop.
We specifically pick the parent, because it is guaranteed to be able to contain children (and ::before), and to avoid conflict
with any potential ::before on the fullscreen element itself.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesFullScreenParentPseudoClass):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/css/fullscreen.css:
(:-webkit-full-screen-parent::before):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:

Canonical link: https://commits.webkit.org/256226@main
